src/mono/System.Private.CoreLib/src/System/Runtime/DependentHandle.cs

```diff
@@ -0,0 +1,170 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Runtime
+{
+    // The ephemeron type is checked in the Mono runtime in "object.c" as follows:
+    //
+    // m_class_get_image (klass) == mono_defaults.corlib && !strcmp ("Ephemeron", m_class_get_name (klass))
+    //
+    // As such, the namespace it belongs to in the managed runtime doesn't matter: the VM will just check
+    // that the type name matches, and that the type is in fact defined in corelib. This means we can
+    // just move it to System.Runtime without worrying about it being properly managed in the VM.
+    // Additionally, the type is defined in "sgen-mono.c" as follows:
+    //
+    // typedef struct {
+    //     GCObject* key;
+    //     GCObject* value;
+    // } Ephemeron;
+    //
+    // So as long as the layout matches the one of the type defined in C# (which it does), we're also free
+    // to rename the fields to better follow the naming guidelines, and the VM will have no issues with it.
+    internal struct Ephemeron
+    {
+        public object? Key;
+        public object? Value;
+    }
+
+    //
+    // Instead of dependent handles, mono uses arrays of Ephemeron objects.
+    //
+    public struct DependentHandle : IDisposable
+    {
+        private Ephemeron[]? _data;
+
+        public DependentHandle(object? target, object? dependent)
+        {
+            _data = new Ephemeron[1];
+            _data[0].Key = target;
+            _data[0].Value = dependent;
+            GC.register_ephemeron_array(_data);
+        }
+
+        public bool IsAllocated => _data is not null;
+
+        public object? Target
+        {
+            get => GetTarget();
+            set => SetTarget(value);
+        }
+
+        public object? Dependent
+        {
+            get => GetDependent();
+            set => SetDependent(value);
+        }
+
+        public (object? Target, object? Dependent) GetTargetAndDependent()
+        {
+            if (_data is not Ephemeron[] data)
```

Comment:
Should this be just a simple null check `if (_data == null)` instead of cast ? I do not see how this can become something else than `Ephemeron[]`.

Reply:
That's not actually a cast (given that the type matches the one of the field, and that the type itself cannot be derived from), the type there is just used to act as a local declaration, so that further accesses to the fields are skipped. The C# compiler will rewrite that as just a local declaration, one field access and then a `null` check as you mentioned. Here's a [sharplab.io repro](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABAzAAgM4AuUArmATgMICwAUAN605M4AOUAlgG4CGBMOAMwA2EXgG0AujgD6AiBADcjZsqbZBI3jgCyACgCUqnAxrMzOdgJy7Z8i3hwA7COWGiCkwfMOnzTE35+GADsOAAMSr5+AL5GZnHMIV4QYmESkWaxNNFAA==). If this style of syntax isn't desired and you'd like me to rewrite it more explicitly, just let me know! 🙂

Reply:
Yes, I think the code would be easier to understand if it make it clear that this is a null check.
```
Ephemeron[] data = _data;
if (data is null)
    throw...
...
```

Reply:
Sure thing! Updated the code style in 418044c17223f7343d621d7822ab701bce6ff917 👍